### PR TITLE
onWillSaveTextDocument event

### PR DESCRIPTION
**Pull request by jrieken**
This PR implements #239 by adding a `onWillSaveTextDocument` which allows for pre-save modifications of a document. The API around that is still in discussion.

### CODE REVIEW

**Comment by jrieken** [comment]
@bpasero Please look at this commit: https://github.com/Microsoft/vscode/pull/12328/commits/7b9f3516f439779ff3c4265b7beb37abcb8f67ab

**Comment by kube** [comment]
I know that you're already working on this, but I wanna expose an issue I have, and I think I will still have after this feature addition.

I have a little extension that adds me a custom header at top of my code.
This header is just composed by multiple comment lines, but left padded by spaces, and it adds an empty line at the top of my file.

https://github.com/kube/vscode-kube-header/blob/master/src/extension.ts

It's pretty simple, but my issue is that code formatters (like Go-format or TypeScript formatter) removes spaces and/or empty lines.
So my extension replaces a broken header by a new one.

I added a hook on save (as you can see in link), to refresh the header.

But, I really need to have this to be executed AFTER the Go-extension formatter. (Or it would be useless)

Would it be possible to add an `order` or `priority` parameter, or dependencies, or something to permit to be sure that a hook is executed AFTER or BEFORE some others.

File: src/vs/workbench/parts/files/common/editors/textFileEditorModel.ts [hunk 1]
```diff
@@ -441,66 +441,81 @@ export class TextFileEditorModel extends BaseTextEditorModel implements ITextFil
...
+				this.blockModelContentChange = true;
+				return TextFileEditorModel.saveParticipant.participate(this, { isAutoSaved });
+			}).then(() => {
 				this.blockModelContentChange = false;
-			}
-			versionId = this.versionId;
+				return this.versionId;
+			}, err => {
+				this.blockModelContentChange = false;
+				return TPromise.wrapError(err);
```
Comment (bpasero) [inline src/vs/workbench/parts/files/common/editors/textFileEditorModel.ts]: @jrieken this is a semantical change from how it worked before where a save participant error would never make the save operation fail. maybe we should ignore the possible errors from save participants (especially since they can now come from extensions) and always allow the save to take place.

File: src/vs/workbench/parts/files/common/editors/textFileEditorModel.ts [hunk 2]
```diff
@@ -441,66 +441,81 @@ export class TextFileEditorModel extends BaseTextEditorModel implements ITextFil
...
+			});
 		}
 
-		// Clear error flag since we are trying to save again
-		this.inErrorMode = false;
+		this.mapPendingSaveToVersionId[versionId] = saveParticipantPromise.then(newVersionId => {
```
Comment (bpasero) [inline src/vs/workbench/parts/files/common/editors/textFileEditorModel.ts]: @jrieken is there a reason why you chose to store the save participant into `mapPendingSaveToVersionId`, then delete it and later store it in `mapPendingSaveToVersionId` again for the `updateContent` call?

Easier to read and understand, I suggest to store the save participant promise into `mapPendingSaveToVersionId` as you do now, but do not delete it right after. Rather let the promise from `this.fileService.updateContent` also contribute to the overall pending save. 

In other words, a pending save is the sum of running save participants and saving to disk.

File: src/vs/workbench/parts/files/common/editors/textFileEditorModel.ts [hunk 3]
```diff
@@ -441,66 +441,81 @@ export class TextFileEditorModel extends BaseTextEditorModel implements ITextFil
...
-				this.setDirty(false);
-			} else {
-				diag(`doSave(${versionId}) - not setting dirty to false because versionId did change meanwhile`, this.resource, new Date());
-			}
+			// Remember when this model was saved last
+			this.lastSaveAttemptTime = Date.now();
+
+			// Save to Disk
+			diag(`doSave(${versionId}) - before updateContent()`, this.resource, new Date());
+			this.mapPendingSaveToVersionId[versionId] = this.fileService.updateContent(this.versionOnDiskStat.resource, this.getValue(), {
```
Comment (bpasero) [inline src/vs/workbench/parts/files/common/editors/textFileEditorModel.ts]: @jrieken as per my comment above I would here just return the promise and not add it to the map, since it is already added from the save participant.

**Comment by bpasero** [comment]
@jrieken done. still unclear to me is how to handle the case of save participants running and the user continue to type into the file. it seems to me that modifications from save participants could kick in anytime while the user is typing. this might be OK if we have a participant that adds imports to the top of the file but maybe less cool, if an edit happens on the same line.
